text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+CLI (the REPL and any command connecting to the server) will use TLS to talk to
+the EdgeDB instances by default. They should advertise the expected ALPN
+protocol in the TLS handshake, and be ready to handle any negotiation result,
+depending on the conclusion we'll make for previous questions.
+
+For compatibility considerations, the client is recommended to provide a
+temporary option to connect to old EdgeDB servers without TLS. However, this
+option should come with a warning and should not be default.
+
+For certificate verification, this was discussed previously. To recap, in order
+to make the client code simple and consistent, the client should either check
+the well-known EdgeDB location for the flag of self-signed certificate in order
+to trust the certificate, or discover the local CA and trust the CA certificate.
+The alternative is to enforce the user to explicitly turn on a switch to allow
+self-signed certificates, but this depends on the result of the discussion.
+
+
+Compatibility & Development
+---------------------------
+
+We mentioned that the new client is recommended to have an option (with
+warnings) to connect to old EdgeDB servers without TLS. In the other hand, we
+won't allow old EdgeDB clients connecting to newer EdgeDB servers with TLS
+enabled. On a TLS handshake failure, the EdgeDB server will return a error in
+clear text, indicating that the client must be upgraded to use TLS.
+
+However, the core EdgeDB server (``edb server``) will provide a hidden option
+``--no-tls`` to run the server in non-TLS compatible mode for development and
+testing only. This option is not available in the EdgeDB CLI
+(``edgedb server``). Without ALPN, the server protocol for ``--no-tls`` will be
+exactly the same as it is now, which is the "<default>" HTTP protocol in the
+previous table, plus the first magic-byte check for the binary protocol.
+
+Note: if we decided to drop the 2 different sets of URI structures, we shouldn't
+add the ``--no-tls`` option too.
+
+Temporarily, the EdgeDB server test suite could run on servers with
+``--no-tls``. But eventually the test suite must match how the server is most
+commonly used, that will be with TLS. So the proposal was to add a new
+subcommand ``edb gen-certs`` to generate self-signed certificate for local
+development and running tests::
+
+    edb gen-certs
+
+    Generate (and update) self-signed certificate for development and testing.
+
+        -D, --data-dir  path to the EdgeDB data directory where the generated
+                        certificate and key files will be stored
+
+But this doesn't work for ``edb server --temp-dir`` or temporary servers
+started in the test suite. Therefore the new proposal is to just generate new
+self-signed certificate while the server is starting in devmode or testmode if
+no exising certificate is found or given. For that, we'd need either the
+OpenSSL library as a new development dependency, or do it through a Rust
```

Comment:
I think we're going to embed cert generation in CLI anyway. So let's just use the CLI.

Codewise cert generation is not that big, we already do that in tests: https://github.com/edgedb/edgedb-cli/blob/master/tests/certs.rs (I'm not sure if binary size of that is big, but I don't think so).